runtime/caml/eventring.h

```diff
@@ -0,0 +1,296 @@
+extern eventring_error
+caml_eventring_create_cursor(const char_os* eventring_path, int pid,
+                             struct caml_eventring_cursor **cursor_res);
+
+/* Set the runtime_begin event callback on the cursor */
+extern void caml_eventring_set_runtime_begin(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, uint64_t timestamp,
+             ev_runtime_phase phase));
+
+/* Set the runtime_end event callback on the cursor */
+extern void caml_eventring_set_runtime_end(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, uint64_t timestamp,
+             ev_runtime_phase phase));
+
+/* Set the runtime_counter event callback on the cursor */
+extern void caml_eventring_set_runtime_counter(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, uint64_t timestamp,
+             ev_runtime_counter counter, uint64_t val));
+
+/* Set the alloc event callback on the cursor */
+extern void
+caml_eventring_set_runtime_alloc(struct caml_eventring_cursor *cursor,
+                                 int (*f)(int domain_id, void *callback_data,
+                                          uint64_t timestamp, uint64_t *sz));
+
+/* Set the lifecycle event callback on the cursor */
+extern void caml_eventring_set_lifecycle(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, int64_t timestamp,
+              ev_lifecycle lifecycle, int64_t data));
+
+/* Set the lost events callback on the cursor */
+extern void caml_eventring_set_lost_events(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, int lost_words));
+
+/* frees a cursor obtained from caml_eventring_creator_cursor */
+extern void
+caml_eventring_free_cursor(struct caml_eventring_cursor *cursor);
+
+/* polls the eventring pointed to by [cursor] and calls the appropriate callback
+    for each new event up to at most [max_events] times.
+
+    Returns the number of events consumed in [events_consumed], if set.
+
+    0 for [max_events] indicates no limit to the number of callbacks. */
+CAMLextern eventring_error caml_eventring_read_poll(
+    struct caml_eventring_cursor *cursor,
+    void *callback_data,
+    uintnat max_events, uintnat *events_consumed);
+
+/* OCaml API for reading from the eventring. Documented in eventring.ml */
+extern value caml_eventring_create_cursor_ml(value path_pid);
+extern value caml_eventring_free_cursor_ml(value wrapped_cursor);
+extern value caml_eventring_read_poll_ml(value wrapped_cursor,
+                                              value callbacks,
+                                              value max_events_option);
+
+#ifdef CAML_INTERNALS
+
+struct eventring_buffer_header {
+  atomic_uint_fast64_t ring_head;
+  atomic_uint_fast64_t ring_tail;
+  uint64_t padding[6]; /* Padding so headers don't share cache lines */
```

Comment:
I think you need to make this [8] rather than [6] in order to avoid false sharing here. I understand that 6 is enough padding to get to cache line size, but what if the first byte in your eventring_buffer_header struct isn't aligned with the beginning of a cacheline?

For example suppose the header starts 8 bytes before the end of a cacheline? Then the `ring_head` of domain N would be on the same cacheline as the `ring_tail` and `padding` of domain N - 1.

Reply:
Yes, good point.

Reply:
Fixed.